### PR TITLE
Add KMS key creation and conditional policy for SNS + Other Services

### PR DESCRIPTION
**Pull request by kristianiliev1**
## Description
Added KMS Key creation. Added policy for it. Added a variable - if set, will create CMK key, if not - will use AWS default keys.

## Related Issues
https://ontotext.atlassian.net/browse/GDB-9961

## Checklist

- [x] I have tested these changes thoroughly.
- [x] My code follows the project's coding style.
- [x] I have added appropriate comments to my code, especially in complex areas.
- [x] All new and existing tests passed locally.

### CODE REVIEW

**Comment by simeonzhekofff** [comment]
1. Use [ GDB-9961 ] without interval to link the MR to the ticket, there is no need to paste the whole link.
2. Create one MR with multiple commits for the different services. ( I see that your MR Is named add kms key creation and conditional policy for SNS, so I assume that you will create different merge requests per "service"). That is not mandatory, instead you can create one MR and push multiple commits named for example: Added CMK support for SNS, Added CMK support for S3, Added CMK support for LB and etc.)

File: modules/monitoring/availability_tests.tf [hunk 1]
```diff
@@ -5,7 +5,9 @@
 resource "aws_sns_topic" "graphdb_route53_sns_topic" {
   provider          = aws.useast1
   name              = "${var.resource_name_prefix}-route53-sns-notifications"
-  kms_master_key_id = "alias/aws/sns"
+  kms_master_key_id = var.enable_cmk ? aws_kms_key.cmk[0].arn : "alias/aws/sns"
+
+
```
Comment (simeonzhekofff) [inline modules/monitoring/availability_tests.tf:10]: Remove excessive empty lines

File: modules/monitoring/cmk.tf [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Creates/manages KMS CMK
+resource "aws_kms_key" "cmk" {
+  count = var.enable_cmk ? 1 : 0
+
+  #description             = var.description
```
Comment (simeonzhekofff) [inline modules/monitoring/cmk.tf:5]: Why this variables is commented?

File: modules/monitoring/cmk.tf [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  customer_master_key_spec = var.key_spec
+  is_enabled               = var.enabled
+  enable_key_rotation      = var.rotation_enabled
+  #tags                    = var.tags
```
Comment (simeonzhekofff) [inline modules/monitoring/cmk.tf:9]: Why this variables is commented?

File: modules/monitoring/cmk.tf [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  deletion_window_in_days = 30
+
+  policy = jsonencode({
+    "Version" : "2012-10-17",
+    "Id" : "key-default-1",
```
Comment (simeonzhekofff) [inline modules/monitoring/cmk.tf:14]: Are you sure that this ID should be named key-default-1?

File: modules/monitoring/cmk.tf [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+          "AWS" : "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
+        },
+        "Action" : "kms:*",
+        "Resource" : "*"
+      },
+      {
+        "Sid" : "Allow access for Key Administrators",
+        "Effect" : "Allow",
+        "Principal" : {
+          #Use 'var.sns_key_admin_arn' if available and root if not provided
```
Comment (simeonzhekofff) [inline modules/monitoring/cmk.tf:29]: leave empty space between the # and the Use word

File: modules/monitoring/cmk.tf [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+          "kms:Decrypt",
+          "kms:ReEncrypt*",
+          "kms:GenerateDataKey*",
+          "kms:DescribeKey"
+        ],
+        "Resource" : "*"
+      }
+    ]
+  })
+
```
Comment (simeonzhekofff) [inline modules/monitoring/cmk.tf:53]: empty line not needed here.

File: modules/monitoring/cmk.tf [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+# Add an alias to the key
+resource "aws_kms_alias" "cmk_alias" {
+  count = var.enable_cmk ? 1 : 0
+
+  name          = "alias/${var.cmk_key_alias}"
+  target_key_id = aws_kms_key.cmk[0].key_id
+}
+
+
```
Comment (simeonzhekofff) [inline modules/monitoring/cmk.tf:64]: remove excessive empty lines

File: modules/monitoring/outputs.tf [hunk 8]
```diff
@@ -0,0 +1,7 @@
+output "cmk_arn" {
+  value = var.enable_cmk && length(aws_kms_key.cmk) > 0 ? aws_kms_key.cmk[0].arn : null
+}
+
+output "cmk_alias_arn" {
+  value = var.enable_cmk && length(aws_kms_alias.cmk_alias) > 0 ? aws_kms_alias.cmk_alias[0].arn : null
+}
```
Comment (simeonzhekofff) [inline modules/monitoring/outputs.tf:7]: Shouldn't we add a description?

File: modules/monitoring/sns.tf [hunk 9]
```diff
@@ -1,8 +1,8 @@
 # SNS Topic
-
```
Comment (simeonzhekofff) [inline modules/monitoring/sns.tf:2]: Why did you remove the empty line?

File: modules/monitoring/sns.tf [hunk 10]
```diff
@@ -1,8 +1,8 @@
...
 resource "aws_sns_topic" "graphdb_sns_topic" {
   name              = "${var.resource_name_prefix}-graphdb-notifications"
-  kms_master_key_id = "alias/aws/sns"
+  kms_master_key_id = var.enable_cmk ? aws_kms_key.cmk[0].arn : "alias/aws/sns"
+
```
Comment (simeonzhekofff) [inline modules/monitoring/sns.tf:5]: Remove the excessive empty line

File: modules/monitoring/variables.tf [hunk 11]
```diff
@@ -119,3 +119,53 @@ variable "route53_availability_check_region" {
   description = "Define route53 health check region"
   type        = string
 }
+
```
Comment (simeonzhekofff) [inline modules/monitoring/variables.tf:122]: Remove the excessive empty lines

File: modules/monitoring/variables.tf [hunk 12]
```diff
@@ -119,3 +119,53 @@ variable "route53_availability_check_region" {
...
+
+#KMS CMK VARS:
```
Comment (simeonzhekofff) [inline modules/monitoring/variables.tf:124]: Leave space between the # and the KMS word

File: modules/monitoring/variables.tf [hunk 13]
```diff
@@ -119,3 +119,53 @@ variable "route53_availability_check_region" {
...
+
+variable "tags" {
+  description = "A map of tags to assign to the resources."
+  type        = map(string)
+  default     = {}
+}
+
+
```
Comment (simeonzhekofff) [inline modules/monitoring/variables.tf:132]: Remove excessive empty lines

File: modules/monitoring/variables.tf [hunk 14]
```diff
@@ -119,3 +119,53 @@ variable "route53_availability_check_region" {
...
+variable "kms_master_key_id" {
+  description = "ARN of the Customer Managed Key (CMK)."
+  type        = string
+  default     = ""
+}
+
+variable "cmk_description" {
```
Comment (simeonzhekofff) [inline modules/monitoring/variables.tf:139]: Where are the type and the description parameters?

Comment (simeonzhekofff) [inline modules/monitoring/variables.tf:139]: Where are the the description and the default parameters?

File: modules/monitoring/variables.tf [hunk 15]
```diff
@@ -119,3 +119,53 @@ variable "route53_availability_check_region" {
...
+  default = "KMS Key to encrypt SNS"
+}
+
+variable "key_spec" {
+  default = "SYMMETRIC_DEFAULT"
+}
+
+variable "enabled" {
```
Comment (simeonzhekofff) [inline modules/monitoring/variables.tf:147]: can we switch the name to something that is giving more information what will be enabled?

File: variables.tf [hunk 16]
```diff
@@ -463,3 +463,39 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
   type        = bool
   default     = false
 }
+
+#cmk building in progress
```
Comment (simeonzhekofff) [inline variables.tf:467]: Leave empty space between the CMK and the #. Also CMK should be with capital letters.

File: variables.tf [hunk 17]
```diff
@@ -463,3 +463,39 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+  default     = false
+}
+
+variable "kms_master_key_id" {
+  description = "ARN of the Customer Managed Key (CMK)."
+  type        = string
+  default     = ""
+}
+
+variable "policy" {
```
Comment (simeonzhekofff) [inline variables.tf:480]: Can we name it cmk_policy for example?

File: variables.tf [hunk 18]
```diff
@@ -463,3 +463,39 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+
+variable "tags" {
+  description = "A map of tags to assign to the resources."
+  type        = map(string)
+  default     = {}
+}
+
+variable "cmk_description" {
+  type    = string
+  default = "KMS Key to encrypt SNS"
```
Comment (simeonzhekofff) [inline variables.tf:494]: You don't have a description parameter.

File: variables.tf [hunk 19]
```diff
@@ -463,3 +463,39 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+}
+
+variable "sns_key_admin_arn" {
+  description = "ARN of the role or user who will have administrative access to the SNS KMS key"
```
Comment (simeonzhekofff) [inline variables.tf:498]: I don't think that's the best description.

File: modules/monitoring/cmk.tf (hunk 4 above)
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:14]: changed to 'kms-key-policy-access-control'

File: modules/monitoring/cmk.tf (hunk 5 above)
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:29]: done.

File: modules/monitoring/cmk.tf (hunk 6 above)
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:53]: fixed

File: modules/monitoring/cmk.tf (hunk 7 above)
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:64]: removed

File: modules/monitoring/outputs.tf (hunk 8 above)
Comment (kristianiliev1) [reply modules/monitoring/outputs.tf:7]: you're right. I'll add

File: modules/monitoring/sns.tf (hunk 9 above)
Comment (kristianiliev1) [reply modules/monitoring/sns.tf:2]: Saw an empty line and removed :D Brought it back

File: modules/monitoring/variables.tf (hunk 14 above)
Comment (kristianiliev1) [reply modules/monitoring/variables.tf:139]: Missing. Will add descriptions. Ty

File: modules/monitoring/variables.tf (hunk 15 above)
Comment (kristianiliev1) [reply modules/monitoring/variables.tf:147]: renamed to key_enabled and added description

File: variables.tf (hunk 16 above)
Comment (kristianiliev1) [reply variables.tf:467]: No, thank you for noticing. I left that one for me to add more variables later.

File: variables.tf (hunk 17 above)
Comment (kristianiliev1) [reply variables.tf:480]: changed to sns_cmk_policy. It is currently not being used anyway, since it is mostly for if one wants to add their own policy. I guess I could remove it also?

File: variables.tf (hunk 18 above)
Comment (kristianiliev1) [reply variables.tf:494]: This is the Variable for the description of the Key. I will make it more clear.

File: variables.tf (hunk 19 above)
Comment (kristianiliev1) [reply variables.tf:498]: Changed to - ARN of the role or user granted administrative access to the SNS KMS key

File: modules/monitoring/cmk.tf (hunk 2 above)
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:5]: fixed

File: modules/monitoring/cmk.tf [hunk 20]
```diff
@@ -0,0 +1,63 @@
+# Creates/manages KMS CMK
+resource "aws_kms_key" "cmk" {
+  count = var.enable_cmk ? 1 : 0
+
+  description              = var.cmk_description
+  customer_master_key_spec = var.key_spec
+  is_enabled               = var.key_enabled
+  enable_key_rotation      = var.rotation_enabled
+  tags                     = var.tags
+  deletion_window_in_days  = 30
```
Comment (viktor-ribchev) [inline modules/monitoring/cmk.tf:10]: make `deletion_window_in_days` a variable

File: modules/monitoring/variables.tf [hunk 21]
```diff
@@ -119,3 +119,54 @@ variable "route53_availability_check_region" {
...
+  type        = string
+  default     = ""
+}
+
+variable "cmk_description" {
+  description = "Description of the Key to be created"
+  default     = "KMS Key to encrypt SNS"
+}
+
+variable "key_spec" {
```
Comment (viktor-ribchev) [inline modules/monitoring/variables.tf:142]: add missing types

**Comment by viktor-ribchev** [comment]
Am I missing something or this is incomplete, where is the CMS setting for S3, EBS,  parameter store?

File: variables.tf [hunk 22]
```diff
@@ -469,3 +469,47 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+  default     = ""
+}
+
+variable "sns_cmk_policy" {
+  description = "The IAM policy JSON document."
+  type        = string
+  default     = ""
+}
+
+variable "tags" {
```
Comment (mihailradkov) [inline variables.tf:493]: This is not used anywhere (because it's not passed in the main.tf). Why do you need this? We already have `common_tags` that are used in provider.tf, effectively adding them to every resource automatically.

File: variables.tf [hunk 23]
```diff
@@ -469,3 +469,47 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
   type        = bool
   default     = false
 }
+
+# KMS CMK
+
+variable "enable_cmk" {
+  description = "Enable CMK for encryption. If false, use AWS managed key."
```
Comment (mihailradkov) [inline variables.tf:476]: I would expand the CMK abbreviation in the description.

Comment (mihailradkov) [inline variables.tf:476]: key or keys?

File: variables.tf [hunk 24]
```diff
@@ -469,3 +469,47 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+}
+
+variable "kms_master_key_id" {
+  description = "ARN of the Customer Managed Key (CMK)."
+  type        = string
+  default     = ""
+}
+
+variable "sns_cmk_policy" {
+  description = "The IAM policy JSON document."
```
Comment (mihailradkov) [inline variables.tf:488]: Could describe a bit more about what it is used for.

File: variables.tf [hunk 25]
```diff
@@ -469,3 +469,47 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+
+# KMS CMK
+
+variable "enable_cmk" {
+  description = "Enable CMK for encryption. If false, use AWS managed key."
+  type        = bool
+  default     = false
+}
+
+variable "kms_master_key_id" {
```
Comment (mihailradkov) [inline variables.tf:481]: I don't see this used anywhere in the monitoring module, do we support passing existing keys or not?

File: variables.tf [hunk 26]
```diff
@@ -469,3 +469,47 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+  description = "ARN of the Customer Managed Key (CMK)."
+  type        = string
+  default     = ""
+}
+
+variable "sns_cmk_policy" {
```
Comment (mihailradkov) [inline variables.tf:487]: Another variable that I think is not even used.

File: variables.tf [hunk 27]
```diff
@@ -469,3 +469,47 @@ variable "graphdb_enable_userdata_scripts_on_reboot" {
...
+}
+
+variable "tags" {
+  description = "A map of tags to assign to the resources."
+  type        = map(string)
+  default     = {}
+}
+
+variable "cmk_description" {
+  description = "Description for the key"
```
Comment (mihailradkov) [inline variables.tf:500]: What key? When does it get used? Add more info.

File: modules/monitoring/cmk.tf [hunk 28]
```diff
@@ -0,0 +1,63 @@
...
+        },
+        "Action" : "kms:*",
+        "Resource" : "*"
+      },
+      {
+        "Sid" : "Allow access for Key Administrators",
+        "Effect" : "Allow",
+        "Principal" : {
+          # Use 'var.sns_key_admin_arn' if available and root if not provided
+          "AWS" : var.sns_key_admin_arn != "" ? var.sns_key_admin_arn : "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
```
Comment (mihailradkov) [inline modules/monitoring/cmk.tf:30]: @simeonzhekofff, is `root` okey here? I'm not familiar with the requirements.

**Review by mihailradkov — CHANGES_REQUESTED** [review]

File: modules/monitoring/cmk.tf (hunk 28 above)
Comment (simeonzhekofff) [reply modules/monitoring/cmk.tf:30]: This parameter defines who is the principal of the key, It shouldn't be root. It's better to be a group for example KeyAdmins or something like that.

File: variables.tf (hunk 26 above)
Comment (kristianiliev1) [reply variables.tf:487]: Perhaps I've been testing something. Thanks for noticing...

File: modules/monitoring/cmk.tf [hunk 29]
```diff
@@ -0,0 +1,129 @@
...
+          "kms:GenerateDataKeyWithoutPlaintext",
+          "kms:DescribeKey"
+        ],
+        "Resource" : "arn:aws:kms:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:key/${aws_kms_key.sns_cmk[0].id}"
+      },
+      {
+        "Sid" : "Allow root user to manage key",
+        "Effect" : "Allow",
+        "Principal" : {
+          "AWS" : "arn:aws:iam::${data.aws_caller_identity.current.account_id}:root"
```
[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:95]: No, the idea is that the Root account have access so if anything goes wrong like it did in the other account, that the root account can fix it.

[the remark this answers is not in the document]
Comment (viktor-ribchev) [reply modules/monitoring/cmk.tf:95]: This doesn't look like a good approach. Imagine deploying this in your department AWS account, something goes wrong and you need to request support from the root account holder. This could be incredibly slow and tedious procedure in some companies.

[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:95]: I agree, however we have had a problem last week with a Key that was created without a policy for some reason and it could not be reached, opened or deleted. That's why I have added this "safety net" just in case. It is only for the root account. I could remove it if you think it is bad, I don't think it will affect the rest of the module, however I have only put it in case it goes wrong agian.

[the remark this answers is not in the document]
Comment (viktor-ribchev) [reply modules/monitoring/cmk.tf:95]: I was rather going in another direction, can't se allow only people with admin roles to access it.

[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:95]: Okay, I'll research for another solution, test it and come back to you.

[the remark this answers is not in the document]
Comment (kristianiliev1) [reply modules/monitoring/cmk.tf:95]: Since there isn't a default admin group/role that we can use, I can create it so that they can (if they want) provide admin group's ARN which will get the full access over the keys and if not provided it will be left empty.

[the remark this answers is not in the document]
Comment (viktor-ribchev) [reply modules/monitoring/cmk.tf:95]: I don't see other options here @simeonzhekofff WDYT?

[the remark this answers is not in the document]
Comment (simeonzhekofff) [reply modules/monitoring/cmk.tf:95]: We can have an admin group that should have the right to assume an IAM role that has permission to take actions on the key.
